core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -233,31 +215,66 @@ public void dispatchBadRequest(
         }
     }
 
-    void dispatchRequest(final RestRequest request, final RestChannel channel, final NodeClient client, ThreadContext threadContext,
-                         final RestHandler handler) throws Exception {
-        if (checkRequestParameters(request, channel) == false) {
-            channel
-                .sendResponse(BytesRestResponse.createSimpleErrorResponse(channel,BAD_REQUEST, "error traces in responses are disabled."));
-        } else {
-            for (String key : headersToCopy) {
-                String httpHeader = request.header(key);
-                if (httpHeader != null) {
-                    threadContext.putHeader(key, httpHeader);
-                }
-            }
+    boolean dispatchRequest(final RestRequest request, final RestChannel channel, final NodeClient client, ThreadContext threadContext,
+                            final RestHandler handler, final PathTrie.TrieMatchingMode trieMatchingMode) throws Exception {
+        final int contentLength = request.hasContent() ? request.content().length() : 0;
 
+        RestChannel responseChannel = channel;
+        if (checkRequestParameters(request, responseChannel) == false) {
+            responseChannel.sendResponse(BytesRestResponse.createSimpleErrorResponse(responseChannel,
+                            BAD_REQUEST, "error traces in responses are disabled."));
+            return true;
+        } else if (contentLength > 0 && hasContentType(request, handler) == false) {
+            sendContentTypeErrorMessage(request, responseChannel);
+            return true;
+        } else if (contentLength > 0 && handler != null && handler.supportsContentStream() &&
+                request.getXContentType() != XContentType.JSON && request.getXContentType() != XContentType.SMILE) {
+            responseChannel.sendResponse(BytesRestResponse.createSimpleErrorResponse(responseChannel,
+                            RestStatus.NOT_ACCEPTABLE, "Content-Type [" + request.getXContentType() +
+                                            "] does not support stream parsing. Use JSON or SMILE instead"));
+            return true;
+        } else {
             if (handler == null) {
-                if (request.method() == RestRequest.Method.OPTIONS) {
-                    // when we have OPTIONS request, simply send OK by default (with the Access Control Origin header which gets automatically added)
-
-                    channel.sendResponse(new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY));
+                /*
+                 * Get the map of matching handlers for a request, for the full set of HTTP methods.
+                 */
+                // TODO: don't explicitly use this TrieMatching method
+                final HashSet<RestRequest.Method> validMethodSet = getValidHandlerMethodSet(request, trieMatchingMode);
```

Comment:
I think we should fix our datastrucuture first and don't make Path trie super complicated and flexible. This should be fixed first before we make this change here.